librosa/feature/utils.py

```diff
@@ -333,3 +333,93 @@ def sync(data, frames, aggregate=None, pad=True):
         start = end
 
     return data_agg
+
+
+@cache
+def chroma_to_tonnetz(chromagram, norm=np.inf):
+    '''Computes the tonal centroid features (tonnetz) from a 12-dimensional
+    chromagram.
+
+    Parameters
+    ----------
+    chromagram : np.ndarray [shape=(12, t)]
+        Normalized energy for each chroma bin at each frame.
+
+    norm : float or None
+        Column-wise normalization.
+        See `librosa.util.normalize` for details.
+
+        If `None`, no normalization is performed.
+
+    Returns
+    -------
+    tonnetz : np.ndarray [shape(6, t)]
+        Tonal centroid features for each frame.
+
+    See Also
+    --------
+    librosa.feature.chroma_cqt
+        Compute a chromagram from a constat-Q transform.
+    librosa.feature.chroma_stft
+        Compute a chromagram from an STFT spectrogram or waveform.
+
+    Examples
+    --------
+    >>> y, sr = librosa.load(librosa.util.example_audio_file())
+    >>> chroma = librosa.feature.chroma_stft(y=y, sr=sr)
+    >>> tonnetz = librosa.feature.chroma_to_tonnetz(chroma)
+    >>> tonnetz
+    array([[ 0.79506535,  0.17179337,  0.30261799, ...,  0.05586615,
+         0.14303219,  0.187812  ],
+       [-0.44264622,  0.41896667,  0.21596437, ...,  0.11552605,
+         0.13349843,  0.92497935],
+       [-1.        ,  0.46310843, -1.        , ...,  0.4240766 ,
+         0.62088668,  0.74145399],
+       [ 0.50550166, -1.        ,  0.33409652, ..., -1.        ,
+        -1.        , -1.        ],
+       [-0.34921771,  0.52859071, -0.14055977, ...,  0.05132072,
+        -0.05028974, -0.22788206],
+       [-0.01232449, -0.19507053, -0.07335521, ..., -0.43501825,
+        -0.49050545, -0.36520804]])
+
+    >>> import matplotlib.pyplot as plt
+    >>> librosa.display.specshow(tonnetz, x_axis='time')
+    >>> plt.colorbar()
+    >>> plt.title('Tonal Centroids (Tonnetz)')
+    >>> plt.tight_layout()
+    '''
+
+    if chromagram.shape[0] != 12:
+        raise ValueError('Tonnetz can only be obtained from 12-dimensional '
+                         'chromagrams.')
+
+    t = chromagram.shape[1]
+    tonnetz = np.zeros((6, t))
+
+    r1 = 1      # Fifths
+    r2 = 1      # Minor
+    r3 = 0.5    # Major
+
+    # Generate Transformation matrix
+    phi = np.zeros((6, 12))
+    for i in range(6):
+        for j in range(12):
+            fun = np.sin if i % 2 == 0 else np.cos
```

Comment:
I think we can also simplify this a bit by replacing the anonymous, alternating function with a `np.pi/2` phase offset for the odd-numbered rows, and just using cosine all the way through.

Similarly, the different scaling fatcors `7/6, 3/2, 2/3` can be stacked up a vector and broadcast over the input.  I think this should make it possible to eliminate all loops from this filter construction.